File: XsquareAndChocolatesBars.py
class XsquareAndChocolatesBars:
    def CountRemainingCandies(self, bar):
        # def CountDifferentCandiesSets(current):
        #     count = 0
        #     while True:
        #         # Stop condition
        #         if current >= len(bar) - 1: return count
        #
        #         # Greedy choice
        #         if bar[current - 1] == bar[current] and bar[current] == bar[current + 1]:
        #             current += 1
        #         else:
        #             count += 1
        #             current += 3

        def CountDifferentCandiesSets(current):
            # Stop condition
            if current >= len(bar) - 1: return 0

            # Greedy choice
            if bar[current - 1] == bar[current] and bar[current] == bar[current + 1]:
                return CountDifferentCandiesSets(current + 1)

            return 1 + CountDifferentCandiesSets(current + 3)

        differentCandiesSetCount = CountDifferentCandiesSets(1)
        return len(bar) - 3 * differentCandiesSetCount

    @staticmethod
    def Work():
        # bar = "CCCCCCCCC"   # Ans: 9
        # bar = "SSSSCSCCC"  # Ans: 3
        # bar = "SSCCSSSCS"  # Ans: 0
        bar = "CCCSCCSSSCSCCSCSSCSCCCSSCCSCCCSCCSSSCCSCCCSCSCCCSSSCCSSSSCSCCCSCSSCSSSCSSSCSCCCSCSCSCSSSCS"  # Ans:

        remainingCandies = XsquareAndChocolatesBars().CountRemainingCandies(bar)
        print(remainingCandies)

File: XsquareAndChocolatesBars.cs
using static System.Console;

namespace Greedy
{
    public class XsquareAndChocolatesBars
    {
        private int CountRemainingCandies(string bar)
        {
            int[] dp = new int[bar.Length + 1];
            int maxGroupsCount = 0;

            CountConsecutiveSets(bar.Length - 3);
            int consecutiveSetsCount = maxGroupsCount;
            return bar.Length - 3 * consecutiveSetsCount;



            void CountConsecutiveSets(int current)
            {
                // Stop condition
                if (current < 0) return;


                // Greedy choice
                // Group w/ unequal pieces: Count 1
                if (bar[current] != bar[current + 1] || bar[current] != bar[current + 2])
                {
                    dp[current] = 1 + dp[current + 3];

                    // Store max no.of groups till current
                    if (dp[current] > maxGroupsCount) maxGroupsCount = dp[current];
                }
                else
                    // Group w/ same pieces: Count 0
                    dp[current] = 0;


                // Find remaining
                CountConsecutiveSets(current - 1);
            }
        }



        internal static void Work()
        {
            //string bar = "CCCCCCCCC";  // Ans: 9
            //string bar = "SSSSCSCCC";  // Ans: 3
            //string bar = "SSCCSSSCS";  // Ans: 0
            //string bar = "CCSSSSC";  // Ans: 1
            string bar = "CCCSCCSSSCSCCSCSSCSCCCSSCCSCCCSCCSSSCCSCCCSCSCCCSSSCCSSSSCSCCCSCSSCSSSCSSSCSCCCSCSCSCSSSCS";  // Ans: 39

            int remainingCandies = new XsquareAndChocolatesBars().CountRemainingCandies(bar);
            WriteLine(remainingCandies);
        }
    }
}